File: src/euler4.rs
use std::cmp::max;

fn length ( n: u32, base: u32) -> u32 {
	let mut length = 0;
	let mut compare = 1;
	loop {
		if compare > n {
			break;
		}
		length += 1;
		compare *= base;
	}
	length
}

fn is_palindrome (n: u32, base: u32) -> bool {
	let len = length(n, base);
	let mid = (len + 1) / 2;
	for i in 0..mid {
		let upper = (n / base.pow(len - i - 1)) % base;
		let lower = (n / base.pow(i)      ) % base;
		if upper != lower {
			return false
		}
	}
	true
}

fn max_palindromic_product ( start: u32, end: u32) -> u32 {
	let mut previous_best = 0;
	for x in start..end {
		for y in start..end {
			let product = x * y;
			if is_palindrome(product, 10) {
				previous_best = max(product, previous_best);
			}
		}
	}
	previous_best
}

/// Find the largest palindrome made from the product of two 3-digit numbers.
fn main() {
	println!("{}", max_palindromic_product(100, 1000));
}

/// A palindromic number reads the same both ways.
/// The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
#[test]
fn test() {
	let expected = 9009;
	let result = max_palindromic_product(10, 100);
	assert_eq!(result, expected);
}

File: c/src/euler4.c
#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <math.h>
#include "integer_maths.c"
/*
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

unsigned max(unsigned a, unsigned b){
	return a < b ? b : a;
}

bool is_palindrome(unsigned N){
	unsigned len = digit_length(N, 10);
	unsigned lower = 10;
	unsigned upper = powi(10, len);
	unsigned low_digit, high_digit;
	for (unsigned i = 0; i < len / 2; i += 1){
		low_digit  = (10 * (N % lower)) / lower;
		high_digit = (10 * (N % upper)) / upper;
		if (low_digit != high_digit){
			return false;
		}
		lower *= 10;
		upper /= 10;
	}
	return true;
}

unsigned largest_palindromic_product_upto(unsigned limit){
	unsigned largest_palindrome = 0;
	for (unsigned i = 1; i < limit; i += 1){
		for (unsigned j = 1; j <= i; j += 1){
			unsigned prod = i*j;
			if (is_palindrome(prod)){
				largest_palindrome = max(prod, largest_palindrome);
			}
		}
	}
	return largest_palindrome;
}

int main(){
	assert(largest_palindromic_product_upto(100) == 9009);
	printf("%u\n", largest_palindromic_product_upto(1000));
	return 0;
}